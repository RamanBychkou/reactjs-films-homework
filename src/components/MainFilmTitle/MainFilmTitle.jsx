import React from 'react';
import shortid from 'shortid';
import style from './MainFilmTitle.scss';
import Star from '../Star/index';
import selectGenre from '../../utils/selectGenre';


export const createMoviesGenreList = (arr, genres, props) => {
  if (arr === undefined) {
    return [];
  }
  return arr.map((c) => {
    let currentGenre;
    genres.some((a) => {
      if (c === a.id) { currentGenre = a.name; return true; }
      return false;
    });
    return <li key={shortid.generate()}><span>{currentGenre}</span></li>;
  });
};


const MainFilmTitle = (props) => {
  const { mainMovie, genres } = props;
  return (
    <div className={style.mainFilmTitle}>
      <h2 className={style.filmTitle}>{mainMovie.title}</h2>
      <div className={style.genres}>
        <ul onClick={selectGenre.bind(props)}>
          {createMoviesGenreList(mainMovie.genre_ids, genres, props)}
        </ul>
      </div>
      <Star rate={mainMovie.vote_average} />
    </div>
  );
};

export default MainFilmTitle;
