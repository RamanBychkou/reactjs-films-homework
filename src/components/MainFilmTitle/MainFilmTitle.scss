.mainFilmTitle {
  color: #fff;
  margin: auto 0 0.7rem;
}

@media screen and (max-width: 46.8rem) {
  .mainFilmTitle {
    margin: 0 0 2rem 0;
  }

  .filmTitle {
    margin-top: 1rem;
  }

  .genres ul {
    flex-flow: column nowrap;

    li {
      margin: 0.3rem;
    }
  }
}

.filmTitle {
  font-size: 1.86rem;
  letter-spacing: 0.09rem;
}

.genres {
  padding: 0.4rem 0 0.3rem 0.1rem;
  font-size: 0.7rem;
  letter-spacing: 0.06rem;
}

.genres ul {
  display: flex;
  justify-content: flex-start;

  li {
    margin-right: 1rem;
    position: relative;

    button {
      background: none;
      border: none;
      padding: 0;
      color: #fff;
      cursor: pointer;

      &:hover {
        color: #47c3f0;
      }
    }
  }

  li:last-child::before {
    content: "";
    border: 1px solid #fff;
    height: 0.6rem;
    left: -9px;
    position: absolute;
  }
}

.mainFilmRate {
  color: #fff;
}

@media screen and (min-width: 82rem) {
  .filmTitle {
    font-size: 2.5rem;
  }

  .genres {
    font-size: 1rem;

    ul li:last-child::before {
      height: 0.9rem;
    }
  }
}
