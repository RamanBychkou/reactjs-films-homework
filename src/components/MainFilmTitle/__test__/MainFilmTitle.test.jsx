import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import MainFilmTitle from '../index';
import { createMoviesGenreList, shooseGenre } from '../MainFilmTitle';

const dataGenres = [
  { id: 28, name: 'Action' },
  { id: 12, name: 'Adventure' },
  { id: 16, name: 'Animation' },
  { id: 35, name: 'Comedy' },
  { id: 80, name: 'Crime' },
  { id: 99, name: 'Documentary' },
  { id: 18, name: 'Drama' },
  { id: 10751, name: 'Family' },
  { id: 14, name: 'Fantasy' },
  { id: 36, name: 'History' },
  { id: 27, name: 'Horror' },
  { id: 10402, name: 'Music' },
  { id: 9648, name: 'Mystery' },
  { id: 10749, name: 'Romance' },
  { id: 878, name: 'Science Fiction' },
  { id: 10770, name: 'TV Movie' },
  { id: 53, name: 'Thriller' },
  { id: 10752, name: 'War' },
  { id: 37, name: 'Western' },
];

test('MainFilmTitle renders correctly', () => {
  const renderer = new ShallowRenderer();
  const mainMovie = {
    adult: false,
    backdrop_path: '/uHEI6v8ApQusjbaRZ8o7WcLYeWb.jpg',
    genre_ids: [27, 878, 18, 53],
    id: 531309,
    original_language: 'en',
    original_title: 'Brightburn',
    overview:
      'What if a child from another world crash-landed on Earth, but instead of becoming a hero to mankind, he proved to be something far more sinister?',
    popularity: 137.601,
    poster_path: '/roslEbKdY0WSgYaB5KXvPKY0bXS.jpg',
    release_date: '2019-05-10',
    title: 'Brightburn',
    video: false,
    vote_average: 7.5,
    vote_count: 2,
  };
  const result = renderer.render(<MainFilmTitle mainMovie={mainMovie} genres={dataGenres} />);
  expect(result).toMatchSnapshot();
});

test('createMoviesGenreList works correctly', () => {
  const testGenresMovies = [53, 10749, 37];
  const movie = {
    mainMovie: {
      runtime: 131,
    },
  };
  expect(createMoviesGenreList(testGenresMovies, dataGenres, movie)).toHaveLength(
    testGenresMovies.length + 1,
  );
});

test('createMoviesGenreList works correctly with integer runtime', () => {
  const testGenresMovies = [53, 10749, 37];
  const movie = {
    mainMovie: {
      runtime: 120,
    },
  };
  expect(createMoviesGenreList(testGenresMovies, dataGenres, movie)).toHaveLength(
    testGenresMovies.length + 1,
  );
});

test('shooseGenre works correctly', () => {
  const elementID = { id: 27 };
  const movie = { name: 'Horror', id: elementID.id };
  const result = shooseGenre(dataGenres, elementID);
  expect(result).toEqual(movie);
});
