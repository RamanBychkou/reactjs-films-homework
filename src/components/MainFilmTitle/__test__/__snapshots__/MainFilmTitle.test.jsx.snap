// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainFilmTitle renders correctly 1`] = `
<div
  className="mainFilmTitle"
>
  <h2
    className="filmTitle"
  >
    Brightburn
  </h2>
  <div
    className="genres"
  >
    <ul />
  </div>
  <Star
    rate={7.5}
  />
</div>
`;
